sormas-app/app/src/main/java/de/symeda/sormas/app/backend/immunization/ImmunizationDao.java

```diff
@@ -1,54 +1,182 @@
+
+		if (!whereStatements.isEmpty()) {
+			Where<Immunization, Long> whereStatement = where.and(whereStatements.size());
+			queryBuilder.setWhere(whereStatement);
+		}
+
+		return queryBuilder;
+	}
+
+	public void deleteImmunizationAndAllDependingEntities(String immunizationUuid) throws SQLException {
+		Immunization immunization = queryUuidWithEmbedded(immunizationUuid);
+
+		// Cancel if not in local database
+		if (immunization == null) {
+			return;
+		}
+
+		// Delete case
+		deleteCascade(immunization);
+	}
+
+	public Immunization build(Person person) {
+		Immunization immunization = super.build();
+		immunization.setPerson(person);
+
+		User user = ConfigProvider.getUser();
+		immunization.setReportingUser(user);
+
+		// Location
+		User currentUser = ConfigProvider.getUser();
+
+		// Set the disease if a default disease is available
+		Disease defaultDisease = DiseaseConfigurationCache.getInstance().getDefaultDisease();
+		if (defaultDisease != null) {
+			immunization.setDisease(defaultDisease);
+		}
+
+		if (UserRole.isPortHealthUser(currentUser.getUserRoles())) {
+			immunization.setResponsibleRegion(currentUser.getRegion());
+			immunization.setResponsibleDistrict(currentUser.getDistrict());
+			immunization.setDisease(Disease.UNDEFINED);
+		} else if (currentUser.getHealthFacility() != null) {
+			immunization.setResponsibleRegion(currentUser.getHealthFacility().getRegion());
+			immunization.setResponsibleDistrict(currentUser.getHealthFacility().getDistrict());
+			immunization.setResponsibleCommunity(currentUser.getHealthFacility().getCommunity());
+		} else {
+			immunization.setResponsibleRegion(currentUser.getRegion());
+			immunization.setResponsibleDistrict(currentUser.getDistrict());
+			immunization.setResponsibleCommunity(currentUser.getCommunity());
+		}
 
-    @Override
-    public String getTableName() {
-        return Immunization.TABLE_NAME;
-    }
+		return immunization;
+	}
 
-    public void deleteImmunizationAndAllDependingEntities(String immunizationUuid) throws SQLException {
-        Immunization immunization = queryUuidWithEmbedded(immunizationUuid);
+	public List<Immunization> getSimilarImmunizations(ImmunizationSimilarityCriteria criteria) {
+		try {
+			QueryBuilder<Immunization, Long> queryBuilder = queryBuilder();
+			QueryBuilder<Person, Long> personQueryBuilder = DatabaseHelper.getPersonDao().queryBuilder();
 
-        // Cancel if not in local database
-        if (immunization == null) {
-            return;
-        }
+			Where<Immunization, Long> where = queryBuilder.where().eq(AbstractDomainObject.SNAPSHOT, false);
+			ImmunizationCriteria immunizationCriteria = criteria.getImmunizationCriteria();
+			where.and().eq(Immunization.DISEASE, immunizationCriteria.getDisease());
+			where.and().eq(Immunization.RESPONSIBLE_REGION + "_id", criteria.getImmunizationCriteria().getResponsibleRegion());
+			if (criteria.getStartDate() != null && criteria.getEndDate() != null) {
```

Comment:
Date filters are incorrect; the system should search for overlapping immunizations. Here's a more detailed definition - each of these scenarios is valid:

* any combination in which either immunization.startDate or immunization.endDate is between criteria.startDate and criteria.endDate
* immunization.startDate == null && ((criteria.startDate == null && criteria.endDate != null) || criteria.startDate < immunization.endDate)
* immunization.endDate == null && ((criteria.endDate == null && criteria.startDate != null) || criteria.endDate > immunization.startDate)

Since you're already in communication with @alexcaruntu-vita, could you please make sure that this is also reflected in the web backend? Thanks!